tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/rules2/Rewriter.scala

```diff
@@ -0,0 +1,14 @@
+package at.forsyte.apalache.tla.bmcmt.rules2
+
+import at.forsyte.apalache.tla.bmcmt.ArenaCell
+import at.forsyte.apalache.tla.lir.TlaEx
+
+/**
+ * @author
+ *   Jure Kukovec
+ */
+trait Rewriter {
+
+  def rewrite(ex: TlaEx)(startingScope: RewriterScope): (RewriterScope, ArenaCell)
```

Comment:
So when rewriting e.g. `\E x \in S: P`, you'll be rewriting the same expression (`P`) under multiple bindings. In that case you'd loop `rewrite(ex)` over different starting scopes. 
Aside from that, I don't really think this sort of thing needs a written elaboration, IIUC there's no performance upside to either of these two approaches.

Reply:
If the usual case is to use it uncurried, then maybe it makes more sense to just curry it in those special cases. That way it will also be clear from context why it's curried. WDYT?